src/tests/remote.rs

```diff
@@ -0,0 +1,150 @@
+use std::str::FromStr;
+
+use git2::{Buf, Error, ErrorClass, Repository};
+
+use crate::git::remote::{get_push_remote, get_upstream, get_upstream_components, set_push_remote, set_upstream, set_upstream_from_ref};
+use crate::tests::helpers::{commit, RepoTestContext};
+
+fn get_head_name(repo: &Repository) -> String {
+    repo.head().unwrap().name().unwrap().into()
+}
+
+fn get_branch_merge(repo: &Repository) -> Result<Buf, Error> {
+    repo.branch_upstream_name(&get_head_name(repo))
+}
+
+fn get_branch_merge_str(repo: &Repository) -> String {
+    get_branch_merge(repo).unwrap().as_str().unwrap().into()
+}
+
+fn get_branch_remote(repo: &Repository) -> Result<Buf, Error> {
+    repo.branch_upstream_remote(&get_head_name(repo))
+}
+
+fn get_branch_remote_str(repo: &Repository) -> String {
+    get_branch_remote(repo).unwrap().as_str().unwrap().into()
+}
+
+fn branch_from_head(repo: &Repository, name: impl AsRef<str>) -> git2::Branch {
+    repo
+        .branch(
+            name.as_ref(),
+            &repo.head().unwrap().peel_to_commit().unwrap(),
+            true,
+        )
+        .unwrap()
+}
+
+#[test]
+fn remove_upstream() {
+    let ctx = RepoTestContext::setup_init();
+
+    let repo = ctx.local_repo;
+    commit(repo.workdir().unwrap(), "file.txt", "content");
+
+    set_upstream(&repo, None).unwrap();
+
+    let e = get_branch_merge(&repo).map(|v| String::from_str(v.as_str().unwrap())).unwrap_err();
+    assert_eq!(e.class(), ErrorClass::Config, "Actual: {}", e);
+
+    let e = get_branch_remote(&repo).map(|v| String::from_str(v.as_str().unwrap())).unwrap_err();
+
+    assert_eq!(e.class(), ErrorClass::Config, "Actual: {}", e);
+}
+
+#[test]
+fn set_new_upstream() {
+    let ctx = RepoTestContext::setup_clone();
+
+    let repo = ctx.local_repo;
+    let remote_name = get_branch_remote_str(&repo);
+    set_upstream(&repo, Some(&format!("{remote_name}/main"))).unwrap();
```

Comment:
```suggestion
    set_upstream(&repo, Some("origin/main")).unwrap();

    assert_eq!(get_branch_remote_str(&repo), "origin");
    assert_eq!(get_branch_merge_str(&repo), "refs/remotes/origin/main");
```

Missed some asserts here, I think it's okay to make some assumptions (remote will be "origin") in tests, makes it more readable :)